src/stores/index.js

```diff
@@ -851,23 +896,38 @@ const Store = Reflux.createStore({
     // Set the connection's app name to the electron app name of Compass.
     connection.appname = electron.remote.app.getName();
 
-    const dataService = new DataService(connection);
+    if (!this.state.isConnecting) {
+      // The connection attempt might have been cancelled while
+      // we were parsing the connection information, so we return here.
+      return;
+    }
 
     try {
-      const runConnect = promisify(dataService.connect.bind(dataService));
-      const connectedDataService = await runConnect();
+      const dataService = new DataService(connection);
+      this.connectingConnectionAttempt = createConnectionAttempt(dataService);
```

Comment:
Why don't we encapsulate the whole connection in the connection attempt?

```
this.connectingConnectionAttempt = createConnectionAttempt(connection);
this.dataService = await this.connectingConnectionAttempt.connect();
 ```
 
 and move all the new Dataservice() stuff etc down in the ConnectionAttempt class?

Reply:
I started to do this, but I found it makes it a bit harder to test just passing the connectionModel to the ConnectionAttempt class. We'd have to be stubbing out a lot of functionality or add some dep injection. I think maybe we leave it like this for now since it's not much of a code change and makes testing a bit easier without having to change up the ConnectionAttempt implementation.